lib/credentials_from_file.c

```diff
@@ -41,46 +41,74 @@ bool write_normal_device_credentials(const char *dev_cred_file,
 		return false;
 	}
 #ifndef NO_PERSISTENT_STORAGE
-	sdow_t sdowriter, *sdow = &sdowriter;
 
-	if (!sdow_init(sdow)) {
+	sdow_t *sdow = sdo_alloc(sizeof(sdow_t));
+	if (!sdow_init(sdow) || !sdo_block_alloc(&sdow->b) || !sdow_encoder_init(sdow)) {
 		LOG(LOG_ERROR, "sdow_init() failed!\n");
-		return false;
+		ret = false;
+		goto end;
 	}
 
 	sdow_next_block(sdow, SDO_DI_SET_CREDENTIALS);
-	sdow_start_array(sdow, 6);
-	sdow_unsigned_int(sdow, ocred->ST);
-	sdow_boolean(sdow, true);
-	sdow_signed_int(sdow, ocred->owner_blk->pv);
-	// See if we can write DeviceCredential.DCHmacSecret here. TO-DO
-	// sdow_byte_string(sdow, ocred->ST, sizeof(ocred->mfg_blk->d));
-	// sdow_text_string(sdow, ocred->mfg_blk->d, sizeof(ocred->mfg_blk->d));
-	sdow_byte_string(sdow, ocred->owner_blk->guid->bytes, GID_SIZE);
-	sdo_rendezvous_list_write(sdow, ocred->owner_blk->rvlst);
-	sdo_hash_write(sdow, ocred->owner_blk->pkh);
-	sdow_end_array(sdow);
-
-	/* Fill sdow buffer */
-
-	if (sdo_blob_write((char *)dev_cred_file, flags, &sdow->b.block[0],
+	if (!sdow_start_array(sdow, 7)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_signed_int(sdow, ocred->ST)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_boolean(sdow, true)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_signed_int(sdow, ocred->owner_blk->pv)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_text_string(sdow, ocred->mfg_blk->d->bytes, ocred->mfg_blk->d->byte_sz)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_byte_string(sdow, ocred->owner_blk->guid->bytes, ocred->owner_blk->guid->byte_sz)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdo_rendezvous_list_write(sdow, ocred->owner_blk->rvlst)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdo_hash_write(sdow, ocred->owner_blk->pkh)) {
+		ret = false;
+		goto end;
+	}
+	if (!sdow_end_array(sdow)) {
+		ret = false;
+		goto end;
+	}
+	size_t encoded_cred_length = 0;
+	if (!sdow_encoded_length(sdow, &encoded_cred_length) || encoded_cred_length <= 0) {
```

Comment:
Does sdow_encoded_length() write a negative number to encoded_cred_length?

Reply:
Updated the usages at all places.